power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -59,20 +60,66 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
-        output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    ComplexValue<sym> const& i_load_gen_bus) {
+    ComplexValue<sym> const i_inj_t = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    std::vector<Idx> sources_acc(sources.size());
+    std::ranges::transform(sources, sources_acc.begin(), [&](Idx const source) { return source; });
+    if (sources_acc.size() == 1) {
+        output.source[sources_acc[0]].i = i_inj_t;
+        output.source[sources_acc[0]].s = output.u[bus_number] * conj(output.source[sources_acc[0]].i);
+    } else if (!sources_acc.empty()) {
+        std::vector<ComplexTensor<sym>> y_ref_acc(sources.size());
+        std::vector<ComplexValue<sym>> i_ref_acc(sources.size());
+        std::ranges::transform(sources, y_ref_acc.begin(), [&](Idx const source) -> ComplexTensor<sym> {
+            return y_bus.math_model_param().source_param[source];
+        });
+        std::ranges::transform(sources, i_ref_acc.begin(), [&](Idx const source) -> ComplexValue<sym> {
+            ComplexValue<sym> const u_ref{input.source[source]};
+            ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
+            return dot(y_ref, u_ref);
+        });
+        ComplexTensor<sym> const y_ref_t = std::accumulate(y_ref_acc.begin(), y_ref_acc.end(), ComplexTensor<sym>{});
+        ComplexTensor<sym> z_ref_t{};
+        if constexpr (is_symmetric_v<sym>) {
+            z_ref_t = 1.0 / y_ref_t;
+        } else {
+            DoubleComplex const s = y_ref_t(0); // Diagonal elements
+            DoubleComplex const m = y_ref_t(1); // Off diagonal elements
+            // z_ref_t is obtained from: y_ref_t -> y_012 -> z_012 -> z_ref_t
+            // s_z, m_z are diagonal and off-diagonal elements of z_ref_t
+            // s_z epsilon (E): 6E + 9E = 15E
+            DoubleComplex const s_z = (2.0 / (3.0 * (s - m))) + (1.0 / (3.0 * (s + (2.0 * m))));
+            // m_z epsilon (E): 9E - 3E = 6E
+            DoubleComplex const m_z = (1.0 / (3.0 * (s + (2.0 * m)))) - (1.0 / (3.0 * (s - m)));
+            z_ref_t = ComplexTensor<sym>{s_z, m_z};
+        }
+        ComplexValue<sym> const i_ref_t = std::accumulate(i_ref_acc.begin(), i_ref_acc.end(), ComplexValue<sym>{});
+        for (size_t i = 0; i < sources.size(); ++i) {
+            Idx const source = sources_acc[i];
+            // The resulting error (from the matrix multiplication) for every u_ref_t element is 54E
+            ComplexValue<sym> const u_ref_t = dot(z_ref_t, i_ref_t);
+            ComplexValue<sym> const u_ref_i{input.source[i]};
+            ComplexValue<sym> delta_u = (u_ref_i - u_ref_t);
+            // This truncation needs further discussion.
+            constexpr double theta = 54.0; // From counting floating point operations involving E.
```

Comment:
one thing I just thought of is that we can also do something like calculate the same function multiple times (3?) but with a small `std::numeric_limits<double>::epsilon()` applied to all input arguments. The standard deviation would then be an estimation of the precision and the mean would be an estimation of the actual value.

The 2 major downsides are:
1. there may be quite a number of input parameters (e.g. an arbitrary symmetric 3-dimensional tensor has 6 inputs. fortunately, the symmetric one with all diagonal elements the same and all off-diagonal elements the same only has 2, so that could help, but only if there's no risk that they differ slightly due to other calculations). We can work around this by doing some arbitrary stuff but it should still be deterministic.
2. depending on the number of input parameters and sources, this can introduce a combinatorically expensive calculation. Fortunately, the expectation is that the amount of sources is small (typically 1 to a couple in the entire grid, let alone on the same node), so this may be premature optimization.

Reply:
I am currently working on how to not use truncation. Since the symmetric tests pass with the current implementation, that seems to indicate that working in the 012 domain may remove the instability. Work in progress.

Reply:
There is no need for truncation anymore after reorganizing the equations cleverly and going to 012 domain for the asymmetric case.